parity-crypto/src/publickey/ecdsa_signature.rs

```diff
@@ -0,0 +1,316 @@
+		self.0[64] += 27;
+		self.0
+	}
+
+	/// Parse bytes as a signature encoded as RSV (V in "Electrum" notation).
+	/// May return empty (invalid) signature if given data has invalid length.
+	pub fn from_electrum(data: &[u8]) -> Self {
+		if data.len() != 65 || data[64] < 27 {
+			// fallback to empty (invalid) signature
+			return Signature::default();
+		}
+
+		let mut sig = [0u8; 65];
+		sig.copy_from_slice(data);
+		sig[64] -= 27;
+		Signature(sig)
+	}
+
+	/// Create a signature object from the sig.
+	pub fn from_rsv(r: &H256, s: &H256, v: u8) -> Self {
+		let mut sig = [0u8; 65];
+		sig[0..32].copy_from_slice(r.as_ref());
+		sig[32..64].copy_from_slice(s.as_ref());
+		sig[64] = v;
+		Signature(sig)
+	}
+
+	/// Check if this is a "low" signature.
+	pub fn is_low_s(&self) -> bool {
+		// "7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF5D576E7357A4501DDFE92F46681B20A0"
+		const MASK: H256 = H256([
+			0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+			0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF,
+			0x5D, 0x57, 0x6E, 0x73, 0x57, 0xA4, 0x50, 0x1D,
+			0xDF, 0xE9, 0x2F, 0x46, 0x68, 0x1B, 0x20, 0xA0,
+		]);
+		H256::from_slice(self.s()) <= MASK
+	}
+
+	/// Check if each component of the signature is in range.
+	pub fn is_valid(&self) -> bool {
+		// "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141"
+		const MASK: H256 = H256([
+			0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+			0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+			0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b,
+			0xbf, 0xd2, 0x5e, 0x8c, 0xd0, 0x36, 0x41, 0x41,
+		]);
+		const ONE: H256 = H256([
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+		]);
+		let r = H256::from_slice(self.r());
+		let s = H256::from_slice(self.s());
+		self.v() <= 1 &&
+			r < MASK && r >= ONE &&
+			s < MASK && s >= ONE
+	}
+}
+
+// manual implementation large arrays don't have trait impls by default.
+// remove when integer generics exist
+impl PartialEq for Signature {
+	fn eq(&self, other: &Self) -> bool {
+		&self.0[..] == &other.0[..]
+	}
+}
+
+// manual implementation required in Rust 1.13+, see `std::cmp::AssertParamIsEq`.
+impl Eq for Signature { }
+
+// also manual for the same reason, but the pretty printing might be useful.
+impl fmt::Debug for Signature {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		f.debug_struct("Signature")
+			.field("r", &self.0[0..32].to_hex::<String>())
+			.field("s", &self.0[32..64].to_hex::<String>())
+			.field("v", &self.0[64..65].to_hex::<String>())
+		.finish()
+	}
+}
+
+impl fmt::Display for Signature {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		write!(f, "{}", self.to_hex::<String>())
+	}
+}
+
+impl FromStr for Signature {
+	type Err = Error;
+
+	fn from_str(s: &str) -> Result<Self, Self::Err> {
+		match s.from_hex::<Vec<u8>>() {
```

Comment:
Is this the only place where we use `rustc-hex`?

Reply:
This and several lines above (in fmt traits implementations)